pyznap/send.py

```diff
@@ -265,114 +262,134 @@ def send_config(config):
+                                 .format(dest_name_log))
                     continue
-                except CalledProcessError as err:
-                    logger.error('Error while opening dest {:s}: \'{:s}\'...'
-                                 .format(dest_name_log, err.stderr.rstrip()))
+            except ValueError as err:
+                logger.error(err)
+                continue
+            except CalledProcessError as err:
+                logger.error('Error while opening dest {:s}: \'{:s}\'...'
+                             .format(dest_name_log, err.stderr.rstrip()))
+                continue
+
+            # Match children on source to children on dest
+            dest_children_names = [child.name.replace(source_name, dest_name) for
+                                   child in source_children]
+            # Send all children to corresponding children on dest
+            for source_fs, dest_name in zip(source_children, dest_children_names):
+                # exclude filesystems from rules
+                if any(fnmatch(source_fs.name, pattern) for pattern in exclude):
+                    logger.debug('Matched {} in exclude rules, not sending...'.format(source_fs))
                     continue
+                # send not excluded filesystems
+                for r in range(retry):
+                    if send_filesystem(source_fs, dest_name, ssh_dest=ssh_dest, raw=raw, resume=resume) == 2:
+                        logger.info('Retrying send in {:d}s (retry {:d} of {:d})...'.format(retry_interval, r+1, retry))
+                        sleep(retry_interval)
+                    else:
+                        break
                 else:
-                    # Match children on source to children on dest
-                    dest_children_names = [child.name.replace(source_name, dest_name) for
-                                           child in source_children]
-                    # Send all children to corresponding children on dest
-                    for source_fs, dest_name in zip(source_children, dest_children_names):
-                        # exclude filesystems from rules
-                        if any(fnmatch(source_fs.name, pattern) for pattern in exclude):
-                            logger.debug('Matched {} in exclude rules, not sending...'.format(source_fs))
-                            continue
-                        # send not excluded filesystems
-                        send_filesystem(source_fs, dest_name, ssh_dest=ssh_dest, raw=raw, resume=resume)
-                finally:
-                    if ssh_dest:
-                        ssh_dest.close()
+                    logger.error('Send was not successful...')
+
+            if ssh_dest:
+                ssh_dest.close()
 
         if ssh_source:
             ssh_source.close()
+
+
+def create_dataset(name, name_log, ssh=None):
```

Comment:
one small bug if you use `zfs.create(name, force=True)` here
if you are non-root, `zfs create -p` will exit 1 and errmsg `filesystem successfully created, but it may only be mounted by root`, even when you have `mount` permission
```
random@gentoo$ zfs allow t                                                                                                                                                                                                                                      
---- Permissions on t ------------------------------------------------
Local+Descendent permissions:
        user random create,destroy,hold,mount,readonly,receive,rename,rollback,send,snapshot,userprop
random@gentoo$ zfs create -p t/backups/remote/new # this command exit with 1                                                                                                                                                                                      
filesystem successfully created, but it may only be mounted by root
random@gentoo$ zfs create -p t/backups/remote/new # this command exit with 0
```
that will stop the `send` if you do not have the parent dataset of your dest, see the test below
```
# here, dest is `t/backups/remote/t`
random@gentoo ~/github/pyznap $ zfs destroy -r t/backups
random@gentoo ~/github/pyznap $ zfs list
NAME              USED  AVAIL     REFER  MOUNTPOINT
t                 947K  6.38G     46.5K  /t
random@gentoo ~/github/pyznap $ pyznap send
Sep 19 10:06:52 INFO: Starting pyznap...
Sep 19 10:06:52 INFO: Sending snapshots...
Sep 19 10:06:52 WARNING: lzop does not exist, continuing without compression...
Sep 19 10:06:52 INFO: Destination t/backups/remote/t does not exist, will create it...
Sep 19 10:06:53 ERROR: Error while creating t/backups/remote/t: Command '['zfs', 'create', '-p', 't/backups/remote/t']' returned non-zero exit status 1....
Sep 19 10:06:53 INFO: Finished successfully...

random@gentoo ~/github/pyznap $ zfs list
NAME              USED  AVAIL     REFER  MOUNTPOINT
t                1007K  6.38G     46.5K  /t
t/backups        46.5K  6.38G     46.5K  /t/backups
random@gentoo ~/github/pyznap $ pyznap send
Sep 19 10:06:58 INFO: Starting pyznap...
Sep 19 10:06:58 INFO: Sending snapshots...
Sep 19 10:06:58 WARNING: lzop does not exist, continuing without compression...
Sep 19 10:06:58 INFO: Destination t/backups/remote/t does not exist, will create it...
Sep 19 10:06:58 ERROR: Error while creating t/backups/remote/t: Command '['zfs', 'create', '-p', 't/backups/remote/t']' returned non-zero exit status 1....
Sep 19 10:06:58 INFO: Finished successfully...

random@gentoo ~/github/pyznap $ zfs list
NAME               USED  AVAIL     REFER  MOUNTPOINT
t                 1021K  6.38G     48.8K  /t
t/backups         93.0K  6.38G     46.5K  /t/backups
t/backups/remote  46.5K  6.38G     46.5K  /t/backups/remote
random@gentoo ~/github/pyznap $ pyznap send
Sep 19 10:07:03 INFO: Starting pyznap...
Sep 19 10:07:03 INFO: Sending snapshots...
Sep 19 10:07:03 WARNING: lzop does not exist, continuing without compression...
Sep 19 10:07:03 INFO: Destination t/backups/remote/t does not exist, will create it...
Sep 19 10:07:03 ERROR: Error while creating t/backups/remote/t: Command '['zfs', 'create', '-p', 't/backups/remote/t']' returned non-zero exit status 1....
Sep 19 10:07:03 INFO: Finished successfully...

random@gentoo ~/github/pyznap $ zfs list
NAME                 USED  AVAIL     REFER  MOUNTPOINT
t                   1.06M  6.38G     51.2K  /t
t/backups            140K  6.38G     46.5K  /t/backups
t/backups/remote    93.0K  6.38G     46.5K  /t/backups/remote
t/backups/remote/t  46.5K  6.38G     46.5K  /t/backups/remote/t
random@gentoo ~/github/pyznap $ pyznap send
Sep 19 10:07:13 INFO: Starting pyznap...
Sep 19 10:07:13 INFO: Sending snapshots...
Sep 19 10:07:13 WARNING: lzop does not exist, continuing without compression...
Sep 19 10:07:13 INFO: No common snapshots on t/backups/remote/t, sending oldest snapshot root@remote:t@0M (~12.6K)...
Sep 19 10:07:14 INFO: Updating t/backups/remote/t with recent snapshot root@remote:t@150M (~150.3M)...
^CSep 19 10:07:15 ERROR: KeyboardInterrupt while sending to t/backups/remote/t...
Sep 19 10:07:15 ERROR: KeyboardInterrupt - exiting gracefully...
```

my origin code about creating ancestors datasets is ugly, but work with this situation: ancestors of dest dataset not existing and you are non-root

Reply:
Hm yes, I didn't consider this, since I always work with root access. This seems like it is a zfs bug (see e.g. https://github.com/openzfs/zfs/issues/8478), since I cannot even create datasets with `zfs create -p -o mountpoint=none test/a/b`, I always get permission problems. I'd rather not have too much code in pyznap handling these special cases, since most of the time you should just create the dataset manually instead of letting pyznap without root access handle it. But what I did now:

```
    try:
        zfs.create(name, ssh=ssh, force=True)
    except CalledProcessError as err:
        message = err.stderr.rstrip()
        if message == "filesystem successfully created, but it may only be mounted by root":
            logger.info('Successfully created {:s}, but cannot mount as non-root...'.format(name_log))
            return 0
        else:
            logger.info('Error while creating {}: \'{:s}\'...'.format(name_log, message))
            return 1
```

So it checks for this error message `filesystem successfully created, but it may only be mounted by root` and will continue with the send if the filesystem was created but not mounted. This means that it can handle creating datasets when the parents exist, but still will fail when the parent not exist. In this case you will have to manually create it.

I also improved the error handling, so it actually shows the error now...